### ISSUE TITLE
Incorrect calculation of a VM memory

### ISSUE THREAD

**Issue by cynepco3hahue** [kubevirt/kubevirt#1100]
I create VM with the specification:
```
...
    resources:
      requests:
        memory: 64Mi
...
```
I assume that VM domain will have 64Mi of memory, but it has 65Mi
```
</metadata>
  <memory unit='KiB'>66560</memory>
  <currentMemory unit='KiB'>66408</currentMemory>
```
When we talk about regular memory, it does not really important, but in case of hugepages, it will fail to start a VM at all.

I think the problem occurs under quantity conversion
```
func QuantityToMegaByte(quantity resource.Quantity) Memory {
	return Memory{
		Value: uint(quantity.ToDec().ScaledValue(6)),
		Unit:  "MB",
	}
}
```

**Comment by fabiand** [comment]
Good catch!

**Comment by fabiand** [comment]
Assigning this to myself to hand it off.

**Comment by vladikr** [comment]
@cynepco3hahue isn't because we add a memory overhead?
Did you request any cpus as well?

**Comment by cynepco3hahue** [comment]
@vladikr We calculate memory overhead for a virt-launcher pod, a VM still must receive the value that we specified under VM spec.
No I did not requested CPU's.

**Comment by mfranczy** [comment]
As @cynepco3hahue mentioned before, there is a problem with
```
Value: uint(quantity.ToDec().ScaledValue(6))
Unit:  "MB",
```
if you will request let's say 67108864 bytes of memory (which is equal to 64MB) then you will receive different value because of scaling and rounding.

To resolve that problem i suggest to change it to:
```
Value: uint64(quantity.AsInt64())
Unit:  "B",
```
and pass memory size in bytes to have exact size which you requested.
It will require change as well in ```Memory.Unit``` type from ```uint``` to ```uint64```

I checked that solution quickly by changing scaling to 0
```
Value: uint(quantity.ToDec().ScaledValue(0))
Unit: "B"
``` 
and it works, below result from my virtual machine:
```
...
<memory unit='KiB'>65536</memory>
<currentMemory unit='KiB'>65536</currentMemory>
<vcpu placement='static'>1</vcpu>
...
```

@cynepco3hahue @fabiand any suggestions or thoughts to this?
If not I will create PR with updated tests

**Comment by cynepco3hahue** [comment]
@mfranczy No objections from my side

**Comment by fabiand** [comment]
Sounds promising, I'd expect some test to the edge cases :)

**Comment by cynepco3hahue** [comment]
@fabiand We have some common test for the converter, we just need to change it, from:
```
var convertedDomain = fmt.Sprintf(`<domain type="%s" 
...
<memory unit="MB">9</memory>
...
```
to
```
var convertedDomain = fmt.Sprintf(`<domain type="%s" 
...
<memory unit="MB">8</memory>
...
```

**Comment by fabiand** [comment]
I wonder if you want to _add_ a new test, to cover even and uneven numbers.

**Comment by mfranczy** [comment]
If you don't mind I think I can add some tests for memory calculations, like checking if you are receiving exactly this what you requested in schema, although I don't have big picture and maybe it doesn't make sense?

**Comment by cynepco3hahue** [comment]
I think it makes sense, a special when you can specify different quantities under memory request(Mi, Gi, M, G,... see https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/ for the full list)

**Comment by mfranczy** [comment]
I have prepared PR for this issue with extended tests for vm memory calculation.

**Comment by fabiand** [comment]
Closed with #1106
